shared/actions/devices.js

```diff
@@ -1,138 +1,192 @@
-        return
+        const error = {errorText: 'No username in removeDevice'}
+        console.warn(error)
+        yield put(({
+          type: Constants.deviceRemoved,
+          payload: error,
+          error: true,
+        }: DeviceRemoved))
       }
-      loginDeprovisionRpc({
-        param: {username, doRevoke: true},
-        callback: error => {
-          if (error) {
-            dispatch({
-              type: Constants.deviceRemoved,
-              payload: error,
-              error: true,
-            })
-          } else {
-            dispatch({
-              type: Constants.deviceRemoved,
-              payload: null,
-              error: false,
-            })
-          }
-
-          if (!error) {
-            dispatch(loadDevices())
-            dispatch(setRevokedSelf(name))
-            dispatch(navigateTo([], loginTab))
-            dispatch(switchTab(loginTab))
-          }
-        },
-      })
-    } else {
-      revokeRevokeDeviceRpc({
+      yield call(loginDeprovisionRpcPromise, {param: {username, doRevoke: true}})
+      yield put(({
+        type: Constants.deviceRemoved,
+        payload: undefined,
+        error: false,
+      }: DeviceRemoved))
+      yield put(loadDevices())
+      yield put(setRevokedSelf(name))
+      yield put(navigateTo([], loginTab))
+      yield put(switchTab(loginTab))
+    } catch (e) {
+      console.warn('Error removing the current device:', e)
+      yield put(({
+        type: Constants.deviceRemoved,
+        payload: e,
+        error: true,
+      }: DeviceRemoved))
+    }
+  } else {
+    // Not the current device.
+    try {
+      yield call(revokeRevokeDeviceRpcPromise, {
         param: {deviceID, force: false},
-        callback: error => {
-          if (error) {
-            dispatch({
-              type: Constants.deviceRemoved,
-              payload: error,
-              error: true,
-            })
-          } else {
-            dispatch({
-              type: Constants.deviceRemoved,
-              payload: null,
-              error: false,
-            })
-          }
+      })
+      yield put(({
+        type: Constants.deviceRemoved,
+        payload: undefined,
+        error: false,
+      }: DeviceRemoved))
+    } catch (e) {
+      console.warn('Error removing a device:', e)
+      yield put(({
+        type: Constants.deviceRemoved,
+        payload: e,
+        error: true,
+      }: DeviceRemoved))
+    }
+  }
+  yield put(loadDevices())
+  yield put(navigateUp(devicesTab, Map({path: 'root'})))
+}
 
-          if (!error) {
-            dispatch(loadDevices())
-            dispatch(navigateUp(devicesTab, Map({path: 'root'})))
-          }
+function _generatePaperKey () {
+  return eventChannel(emit => {
+    loginPaperKeyRpc({
+      incomingCallMap: {
+        'keybase.1.loginUi.promptRevokePaperKeys': (param, response) => {
+          // We only pretend to support this RPC.
+          response.result(false)
```

Comment:
Hm, that introduces ordering into the channel takes, and so I'm worried about what would happen if the order the RPCs are generated by the service flips.  Any ideas?